### ISSUE TITLE
[native-image] should not reach here with grpc-java

### ISSUE THREAD

**Issue by soushin** [oracle/graal#754]
I'm trying to build native image from this project: https://github.com/soushin/grpc-server-on-graalvm  
`grpc-server-on-graalvm` includes plain gRPC Server.   
Is GraalVM support to serving gRPC?

Build log is following:

```
root@53f525ee0086:/usr/local/app/lib# native-image --verbose -H:Name=app -jar app.jar -H:ReflectionConfigurationResources=graal-app.json
StartServer [
/graalvm/bin/java \
-Xbootclasspath/a:/graalvm/jre/lib/boot/graal-sdk.jar:/graalvm/jre/lib/boot/graaljs-scriptengine.jar \
-cp \
/graalvm/jre/lib/svm/builder/pointsto.jar:/graalvm/jre/lib/svm/builder/objectfile.jar:/graalvm/jre/lib/svm/builder/svm.jar:/graalvm/jre/lib/jvmci/jvmci-api.jar:/graalvm/jre/lib/jvmci/graal-management.jar:/graalvm/jre/lib/jvmci/jvmci-hotspot.jar:/graalvm/jre/lib/jvmci/graal.jar \
-server \
-d64 \
-noverify \
-XX:+UnlockExperimentalVMOptions \
-XX:+EnableJVMCI \
-XX:-UseJVMCIClassLoader \
-Dgraal.EagerSnippets=true \
-Xss10m \
-Duser.country=US \
-Duser.language=en \
-Dgraalvm.version=1.0.0-rc7 \
-Dorg.graalvm.version=1.0.0-rc7 \
-Dcom.oracle.graalvm.isaot=true \
-Djvmci.class.path.append=/graalvm/jre/lib/jvmci/graal.jar \
-Xmx3344270948 \
-Xms1g \
com.oracle.svm.hosted.server.NativeImageBuildServer \
-port=0 \
-logFile=/root/.native-image/machine-id-3d6ff2f75c7d3ae927580249a28e7e05/session-id-1/server-id-d70d087cd99ee9332c8fbb61b3468d10cf83f7a0cd6ee1ac3d5635ce79844497df69f41a75e16c694c1f0df4c3df9f81a1ba7eec3a8562fe1ec6349fb4215ac1/server.log
]
Build on Server(pid: 30, port: 34821)*
SendBuildRequest [
-task=com.oracle.svm.hosted.NativeImageGeneratorRunner
-imagecp
/graalvm/jre/lib/svm/builder/pointsto.jar:/graalvm/jre/lib/svm/builder/objectfile.jar:/graalvm/jre/lib/svm/builder/svm.jar:/graalvm/jre/lib/jvmci/jvmci-api.jar:/graalvm/jre/lib/jvmci/graal-management.jar:/graalvm/jre/lib/jvmci/jvmci-hotspot.jar:/graalvm/jre/lib/jvmci/graal.jar:/graalvm/jre/lib/boot/graal-sdk.jar:/graalvm/jre/lib/boot/graaljs-scriptengine.jar:/graalvm/jre/lib/svm/library-support.jar:/usr/local/app/lib:/usr/local/app/lib/app.jar
-H:Path=/usr/local/app/lib
-H:Class=me.soushin.app.Application
-H:Name=app
-H:CLibraryPath=/graalvm/jre/lib/svm/clibraries/linux-amd64
-H:ReflectionConfigurationResources=graal-app.json
]
[app:30]    classlist:  13,002.00 ms
[app:30]        (cap):   1,322.74 ms
Warning: Detected unnecessary RecomputeFieldValue.ArrayBaseOffset com.oracle.svm.core.jdk.Target_java_nio_DirectByteBuffer.arrayBaseOffset substitution field for java.nio.DirectByteBuffer.arrayBaseOffset. The annotated field can be removed. This ArrayBaseOffset computation can be detected automatically. Use option -H:+UnsafeAutomaticSubstitutionsLogLevel=2 to print all automatically detected substitutions.
[app:30]        setup:   3,760.85 ms
[app:30]     analysis:   5,160.15 ms
fatal error: com.oracle.svm.core.util.VMError$HostedError: should not reach here
	at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:62)
	at com.oracle.svm.hosted.SVMHost.dynamicHub(SVMHost.java:201)
	at com.oracle.svm.hosted.analysis.Inflation.checkType(Inflation.java:139)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080)
	at com.oracle.svm.hosted.analysis.Inflation.checkObjectGraph(Inflation.java:130)
	at com.oracle.graal.pointsto.BigBang.checkObjectGraph(BigBang.java:590)
	at com.oracle.graal.pointsto.BigBang.finish(BigBang.java:562)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:691)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:402)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Error: Processing image build request failed
```

To reproduce you need to perform following steps;

1. git clone git@github.com:soushin/grpc-server-on-graalvm.git
2. mvn install `graalvm-ce-1.0.0-rc8` for substitution (see also [netty-native-demo](https://github.com/cstancu/netty-native-demo#setting-up-the-project))
3. cd grpc-server-on-graalvm
4. ./gradlew clean build
5. docker build -t soushin/grpc-server-on-graalvm .

My GraalVM version:

```
root@53f525ee0086:/# java -version
openjdk version "1.8.0_172"
OpenJDK Runtime Environment (build 1.8.0_172-20180625212755.graaluser.jdk8u-src-tar-g-b11)
GraalVM 1.0.0-rc7 (build 25.71-b01-internal-jvmci-0.48, mixed mode)
```

**Comment by cstancu** [comment]
@soushin I am not sure if we support gRPC Server out of the box, there are probably some [limitations](https://github.com/oracle/graal/blob/master/substratevm/LIMITATIONS.md) that you will need to address. However, looking at your stack trace I don't see how `SVMHost.java:201` can throw a `VMError.shouldNotReachHere` since the code at that line is a cast and an assignment: `aType = (AnalysisType) type`. Looking at your steps to reproduce you mention `mvn install graalvm-ce-1.0.0-rc8` but your GraalVM version is `GraalVM 1.0.0-rc7`. Can you please [download](https://github.com/oracle/graal/releases) `graalvm-ce-1.0.0-rc8` and try again?

**Comment by soushin** [comment]
@cstancu Thanks for your reply :)

I"m trying with `graalvm-ce-1.0.0-rc8`, but same errors occured.

```
root@11f9c9a6cfe5:/# java -version
openjdk version "1.8.0_172"
OpenJDK Runtime Environment (build 1.8.0_172-20180625212755.graaluser.jdk8u-src-tar-g-b11)
GraalVM 1.0.0-rc8 (build 25.71-b01-internal-jvmci-0.48, mixed mode)
root@11f9c9a6cfe5:/# cd /usr/local/app/lib
root@11f9c9a6cfe5:/usr/local/app/lib# native-image --verbose -H:Name=app -jar app.jar -H:ReflectionConfigurationResources=graal-app.json
StartServer [
/graalvm/bin/java \
-Xbootclasspath/a:/graalvm/jre/lib/boot/graal-sdk.jar:/graalvm/jre/lib/boot/graaljs-scriptengine.jar \
-cp \
/graalvm/jre/lib/svm/builder/pointsto.jar:/graalvm/jre/lib/svm/builder/objectfile.jar:/graalvm/jre/lib/svm/builder/svm.jar:/graalvm/jre/lib/jvmci/jvmci-api.jar:/graalvm/jre/lib/jvmci/graal-management.jar:/graalvm/jre/lib/jvmci/jvmci-hotspot.jar:/graalvm/jre/lib/jvmci/graal.jar \
-server \
-d64 \
-noverify \
-XX:+UnlockExperimentalVMOptions \
-XX:+EnableJVMCI \
-XX:-UseJVMCICompiler \
-XX:-UseJVMCIClassLoader \
-Dgraal.EagerSnippets=true \
-Xss10m \
-Duser.country=US \
-Duser.language=en \
-Dgraalvm.version=1.0.0-rc8 \
-Dorg.graalvm.version=1.0.0-rc8 \
-Dcom.oracle.graalvm.isaot=true \
-Djvmci.class.path.append=/graalvm/jre/lib/jvmci/graal.jar \
-Xmx3344270948 \
-Xms1g \
com.oracle.svm.hosted.server.NativeImageBuildServer \
-port=0 \
-logFile=/root/.native-image/machine-id-3d6ff2f75c7d3ae927580249a28e7e05/session-id-1/server-id-dfd8c5256727a54ac84aee0bd855462c0d20f087e56a1bdab35b208cc7a54277d6fcc33f17cdc6444c8c94c5c7809c5d5a476551d0551aea4b173c426f58b6f8/server.log
]
Build on Server(pid: 29, port: 45887)*
SendBuildRequest [
-task=com.oracle.svm.hosted.NativeImageGeneratorRunner
-imagecp
/graalvm/jre/lib/svm/builder/pointsto.jar:/graalvm/jre/lib/svm/builder/objectfile.jar:/graalvm/jre/lib/svm/builder/svm.jar:/graalvm/jre/lib/jvmci/jvmci-api.jar:/graalvm/jre/lib/jvmci/graal-management.jar:/graalvm/jre/lib/jvmci/jvmci-hotspot.jar:/graalvm/jre/lib/jvmci/graal.jar:/graalvm/jre/lib/boot/graal-sdk.jar:/graalvm/jre/lib/boot/graaljs-scriptengine.jar:/graalvm/jre/lib/svm/library-support.jar:/usr/local/app/lib:/usr/local/app/lib/app.jar
-H:Path=/usr/local/app/lib
-H:Class=me.soushin.app.Application
-H:Name=app
-H:CLibraryPath=/graalvm/jre/lib/svm/clibraries/linux-amd64
-H:ReflectionConfigurationResources=graal-app.json
]
[app:29]    classlist:   7,535.41 ms
[app:29]        (cap):   1,085.86 ms
[app:29]        setup:   3,072.03 ms
[app:29]     analysis:   5,514.26 ms
fatal error: com.oracle.svm.core.util.VMError$HostedError: should not reach here
	at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:62)
	at com.oracle.svm.hosted.SVMHost.dynamicHub(SVMHost.java:205)
	at com.oracle.svm.hosted.analysis.Inflation.checkType(Inflation.java:139)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080)
	at com.oracle.svm.hosted.analysis.Inflation.checkObjectGraph(Inflation.java:130)
	at com.oracle.graal.pointsto.BigBang.checkObjectGraph(BigBang.java:590)
	at com.oracle.graal.pointsto.BigBang.finish(BigBang.java:562)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:691)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:402)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Error: Processing image build request failed
```

**Comment by cstancu** [comment]
@soushin thanks for trying with rc8. The error is the same, but now the trace has valid line numbers. This problem is caused by trying to build the native image with an incomplete class path, i.e., the app that you are trying to build the image from has some optional library dependencies which you probably don't use. We are currently working on a general solution for these issues and plan to release it in rc9.

**Comment by soushin** [comment]
@cstancu I tried rc10, confirmed valid gRPC server is running!
I added `--allow-incomplete-classpath` new option for building native image and added `--delay-class-initialization-to-runtime` option.  
The code difference is following.

https://github.com/soushin/grpc-server-on-graalvm/pull/1/files

Thanks for great updating! We are going to continue to verify that we are able to introduce or not GraalVM to our project.

Could I close this issue?
